test/SILOptimizer/devirtualize_protocol_composition_two_stores.sil

```diff
@@ -0,0 +1,114 @@
+// This test makes sure that we look at the correct uses with the
+// correct dominance order (store before apply before dealloc).
+// This function will be passed arguments of type B and C for arguments
+// %0 and %1 respectively. We want to make sure that we call B's foo method
+// and not C's foo method.
+sil hidden [noinline] @impl : $@convention(thin) (@guaranteed A & P, @guaranteed A & P) -> Int {
+bb0(%0 : $A & P, %1 : $A & P):
+  %2 = open_existential_ref %0 : $A & P to $@opened("A1F1B526-1463-11EA-932F-ACBC329C418B") A & P
+  %3 = unchecked_ref_cast %1 : $A & P to $@opened("A1F1B526-1463-11EA-932F-ACBC329C418B") A & P
+
+  %5 = alloc_stack $@opened("A1F1B526-1463-11EA-932F-ACBC329C418B") A & P
+  store %2 to %5 : $*@opened("A1F1B526-1463-11EA-932F-ACBC329C418B") A & P
+  
+  // We want to make sure that we picked up on the FIRST store and not the second one.
+  // class C's foo method is named "bar" whereas class B's foo method is named "foo".
+  // We want to make sure that we call a function named "foo" not "bar".
+  // CHECK: [[FN:%[0-9]+]] = function_ref @${{.*}}foo{{.*}} : $@convention(thin) () -> Int
+  %7 = witness_method $@opened("A1F1B526-1463-11EA-932F-ACBC329C418B") A & P, #P.foo!1 : <Self where Self : P> (Self) -> () -> Int, %2 : $@opened("A1F1B526-1463-11EA-932F-ACBC329C418B") A & P : $@convention(witness_method: P) <τ_0_0 where τ_0_0 : P> (@in_guaranteed τ_0_0) -> Int
+  // CHECK: apply [[FN]]
+  %8 = apply %7<@opened("A1F1B526-1463-11EA-932F-ACBC329C418B") A & P>(%5) : $@convention(witness_method: P) <τ_0_0 where τ_0_0 : P> (@in_guaranteed τ_0_0) -> Int
+  
+  store %3 to %5 : $*@opened("A1F1B526-1463-11EA-932F-ACBC329C418B") A & P
+  dealloc_stack %5 : $*@opened("A1F1B526-1463-11EA-932F-ACBC329C418B") A & P
+  return %8 : $Int
+} // end sil function 'impl'
+
+// protocol witness for P.foo() in conformance B
+sil shared [transparent] [serialized] [thunk] @Pfoo : $@convention(witness_method: P) (@in_guaranteed B) -> Int {
+// %0                                             // user: %1
+bb0(%0 : $*B):
+  %1 = load %0 : $*B                              // users: %2, %3
+  %2 = class_method %1 : $B, #B.foo!1 : (B) -> () -> Int, $@convention(method) (@guaranteed B) -> Int // user: %3
+  %3 = apply %2(%1) : $@convention(method) (@guaranteed B) -> Int // user: %4
+  return %3 : $Int                                // id: %4
+} // end sil function 'Pfoo'
+
+// B.foo()
+sil [noinline] @foo : $@convention(method) (@guaranteed B) -> Int {
+// %0                                             // user: %1
+bb0(%0 : $B):
+  debug_value %0 : $B, let, name "self", argno 1  // id: %1
+  %2 = integer_literal $Builtin.Int64, 1          // user: %3
+  %3 = struct $Int (%2 : $Builtin.Int64)          // user: %4
+  return %3 : $Int                                // id: %4
+} // end sil function 'foo'
+
+// test(x:y:)
+sil @test :$@convention(thin) (@guaranteed B, @guaranteed C) -> Int {
```

Comment:
I'm unclear as to why the body of `@test` is included but nothing is checked. Is it just to illustrate how `impl` might be called? You could do that in a comment. Or is the SIL actually necessary for the test? Please clarify with a comment, or just remove the SIL body if it doesn't add anything to the test.

Reply:
Yes, `test` is required because it generates a specialization of `impl`. Without `test`, `impl` isn't optimized (in the way we want) because it doesn't know what parameters it is getting. I'll add some comments to make it more clear what's happening.